numpy/lib/format.py

```diff
@@ -304,27 +305,32 @@ def _write_array_header(fp, d, version=None):
         header.append("'%s': %s, " % (key, repr(value)))
     header.append("}")
     header = "".join(header)
-    # Pad the header with spaces and a final newline such that the magic
-    # string, the header-length short and the header are aligned on a
-    # 16-byte boundary.  Hopefully, some system, possibly memory-mapping,
-    # can take advantage of our premature optimization.
-    current_header_len = MAGIC_LEN + 2 + len(header) + 1  # 1 for the newline
-    topad = 16 - (current_header_len % 16)
-    header = header + ' '*topad + '\n'
     header = asbytes(_filter_header(header))
 
-    hlen = len(header)
-    if hlen < 256*256 and version in (None, (1, 0)):
+    hlen = len(header) + 1 # 1 for newline
+    padlen1 = ARRAY_ALIGN - ((MAGIC_LEN + struct.calcsize('<H') + hlen) % ARRAY_ALIGN)
+    padlen2 = ARRAY_ALIGN - ((MAGIC_LEN + struct.calcsize('<I') + hlen) % ARRAY_ALIGN)
+
+    if hlen + padlen1 < 2**16 and version in (None, (1, 0)):
```

Comment:
@eric-wieser I tried using `iinfo('<H')` as you suggested but it introduces a dependency on another module which can result in a circular import error.  Since these magic numbers existed in much the same way before this PR, I prefer to just leave well enough alone and if magic numbers must be removed from the source code, it can be done on a separate PR, rather than risking introducing breaking changes into this one.

@seberg I have added a comment in 0073949 explaining what 2**16 is doing there.